tools/tpm/policy_sign.c

```diff
@@ -0,0 +1,198 @@
+/* policy_sign.c
+ *
+ * Copyright (C) 2006-2023 wolfSSL Inc.
+ *
+ * This file is part of wolfBoot.
+ *
+ * wolfBoot is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * wolfBoot is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+#include <stdint.h>
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/wolfcrypt/ecc.h>
+#include <wolfssl/wolfcrypt/hash.h>
+#include <wolftpm/tpm2_wrap.h>
+#include <hal/tpm_io.h>
+#include <wolfssl/wolfcrypt/random.h>
+#include <keystore.h>
+#include <tpm.h>
+
+
+static void usage(void)
+{
+    printf("Expected usage:\n");
+    printf("./examples/pcr/policy_sign policy.bin key\n");
+}
+
+static int read_file(const char *path, uint8_t **buf, size_t *size)
+{
+    *buf = NULL;
+    int io_sz;
+    FILE *f;
+
+    f = fopen(path, "rb");
+    if (f == NULL) {
+        printf("Open key file %s failed\n", path);
+        return -1;
+    }
+    fseek(f, 0, SEEK_END);
+    *size = ftell(f);
+    fseek(f, 0, SEEK_SET);
+    *buf = malloc(*size);
+    if (*buf) {
+        io_sz = (int)fread(*buf, 1, *size, f);
+        if (io_sz != (int)*size) {
+            printf("Key file read error!\n");
+            goto exit0;
+        }
+    }
+exit0:
+    fclose(f);
+    return 0;
+}
+
+static int sign_digest(const uint8_t *digest, int digest_sz, uint8_t *signature,
+                       ecc_key *k)
+{
+    int sig_sz = 32;
+    mp_int r, s;
+    WC_RNG rng;
+    int ret;
+
+    ret = wc_InitRng(&rng);
+    if (ret != 0) {
```

Comment:
Please convert this into a single return and make sure mp_clear is called after mp_init. The sign failure may not call mp_clear which could leak memory for some math implementations.